types/src/transaction/webauthn.rs

```diff
@@ -0,0 +1,693 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction::authenticator::AnyPublicKey;
+use anyhow::{anyhow, Result};
+use aptos_crypto::{
+    hash::CryptoHash, p256_ecdsa, signing_message, CryptoMaterialError, HashValue, Signature,
+};
+use passkey_types::{crypto::sha256, webauthn::CollectedClientData, Bytes};
+use serde::{Deserialize, Serialize};
+
+/// Returns the binary concatenation of
+/// 1. [`authenticator_data_bytes`](PartialAuthenticatorAssertionResponse) and
+/// 2. SHA-256 hash of [`client_data_json`](PartialAuthenticatorAssertionResponse),
+///
+/// See <https://www.w3.org/TR/webauthn-3/#sctn-verifying-assertion>
+fn generate_verification_data(authenticator_data_bytes: &[u8], client_data_json: &[u8]) -> Vec<u8> {
+    // Let hash be the result of computing a hash over the clientData using SHA-256.
+    let client_data_json_hash = sha256(client_data_json);
+    // Binary concatenation of authData and hash.
+    // Note: This is compatible with signatures generated by FIDO U2F
+    // authenticators. See §6.1.2 FIDO U2F Signature Format Compatibility
+    // See <https://www.w3.org/TR/webauthn-3/#sctn-fido-u2f-sig-format-compat>
+    authenticator_data_bytes
+        .iter()
+        .chain(client_data_json_hash.iter())
+        .copied()
+        .collect::<Vec<u8>>()
+}
+
+/// This checks that the SHA3-256 of the `signing_message(message)`
+/// is equal to the actual `challenge` from `client_data_json` in `PartialAuthenticatorAssertionResponse`
+fn verify_expected_challenge_from_message_matches_actual<T: Serialize + CryptoHash>(
+    message: &T,
+    actual_challenge: &[u8],
+) -> std::result::Result<bool, CryptoMaterialError> {
+    // Generate signing_message
+    let signing_message_bytes = signing_message(message)?;
```

Comment:
so the signing msg is still bcs-ed bytes of raw txn, right?
Please make a more detailed comments here. I think this line and line 150 are both crucial to understand the whole flow.

Reply:
Sure will add some more detailed comments. Yeah `signing_msg = HASH + BCS(raw_transaction)`